dbms/src/Storages/MergeTree/MergeTreeUniqueIndex.cpp

```diff
@@ -0,0 +1,411 @@
+    traverseAST(expression_ast);
+
+    auto syntax_analyzer_result = SyntaxAnalyzer(context, {}).analyze(
+            expression_ast, index.header.getNamesAndTypesList());
+    actions = ExpressionAnalyzer(expression_ast, syntax_analyzer_result, context).getActions(true);
+}
+
+bool UniqueCondition::alwaysUnknownOrTrue() const
+{
+    return useless;
+}
+
+bool UniqueCondition::mayBeTrueOnGranule(MergeTreeIndexGranulePtr idx_granule) const
+{
+    auto granule = std::dynamic_pointer_cast<MergeTreeUniqueGranule>(idx_granule);
+    if (!granule)
+        throw Exception(
+                "Unique index condition got wrong granule", ErrorCodes::LOGICAL_ERROR);
+
+    if (useless)
+        return true;
+
+    if (index.max_rows && granule->size() > index.max_rows)
+        return true;
+
+    Block result = granule->getElementsBlock();
+    actions->execute(result);
+
+
+    const auto & column = result.getByName(expression_ast->getColumnName()).column;
+
+    for (size_t i = 0; i < column->size(); ++i)
+        if (column->getBool(i))
+            return true;
+
+    return false;
+}
+
+void UniqueCondition::traverseAST(ASTPtr & node) const
+{
+    if (operatorFromAST(node))
+    {
+        auto * func = typeid_cast<ASTFunction *>(&*node);
+        auto & args = typeid_cast<ASTExpressionList &>(*func->arguments).children;
+
+        for (auto & arg : args)
+            traverseAST(arg);
+        return;
+    }
+
+    if (!atomFromAST(node))
+        node = std::make_shared<ASTLiteral>(Field(3)); /// can_be_true=1 can_be_false=1
+}
+
+bool UniqueCondition::atomFromAST(ASTPtr & node) const
+{
+    /// Function, literal or column
+
+    if (typeid_cast<const ASTLiteral *>(node.get()))
+        return true;
+
+    if (const auto * identifier = typeid_cast<const ASTIdentifier *>(node.get()))
+        return key_columns.count(identifier->getColumnName()) != 0;
+
+    if (auto * func = typeid_cast<ASTFunction *>(node.get()))
+    {
+        if (key_columns.count(func->getColumnName()))
+        {
+            /// Function is already calculated.
+            node = std::make_shared<ASTIdentifier>(func->getColumnName());
+            return true;
+        }
+
+        ASTs & args = typeid_cast<ASTExpressionList &>(*func->arguments).children;
+
+        for (auto & arg : args)
+            if (!atomFromAST(arg))
+                return false;
+
+        return true;
+    }
+
+    return false;
+}
+
+bool UniqueCondition::operatorFromAST(ASTPtr & node) const
+{
+    /// Functions AND, OR, NOT. Replace with bit*.
+    auto * func = typeid_cast<ASTFunction *>(&*node);
+    if (!func)
+        return false;
+
+    const ASTs & args = typeid_cast<const ASTExpressionList &>(*func->arguments).children;
+
+    if (func->name == "not")
+    {
+        if (args.size() != 1)
+            return false;
+
+        const auto one = std::make_shared<ASTLiteral>(Field(1));
+        const auto two = std::make_shared<ASTLiteral>(Field(2));
+
+        node = makeASTFunction(
```

Comment:
Would special function for swapping last two bits be better in this case?

Reply:
Yes, but this function would not be usable for anything else.

Reply:
Let's make this function nevertheless.